### DISCUSSION TITLE
Build Output API `routes` unexpected rewriting behavior?

### DISCUSSION THREAD

**Discussion by AaronBeaudoin** [vercel/vercel#8329] [Help] [answered]
Hello,

I am trying to use the Build Output API to deploy my site. I have 2 serverless functions:
- `.vercel/output/functions/default.func`
- `.vercel/output/functions/direct.func`

The first function (`default.func`) is a prerender function, but I don't really think that matters for this question. (_Edit: Turns out I'm wrong on this, see my update at the bottom of the post._)

I would like my routing to work like this:

- If a static file exists, serve that and be done.
- If the path starts with `/direct` send the request to `direct.func`.
- For all other routes, send the request to `default.func`.

Here is my `.vercel/output/config.json`:

```json
{
  "version": 3,
  "routes": [
    { "handle": "filesystem" },
    { "src": "/direct/(.*)", "dest": "/direct" },
    { "src": "/(.*)", "dest": "/default" }
  ]
}
```

The problem is, all paths **not** starting with `/direct` that get sent to `default.func` get their path changed to `/default` inside the function, regardless of what the `src` path is. In other words, `req.url` is always `/default`. What I would actually want is the function to get a `req.url` of whatever the original `src` URL was. This is what happens for my `/direct` function routing rule. This is also what happens in another similar project of mine where I only have a single rewrite. Why does it not work in this case the same as expected?? Here is a screenshot of my function logs:

<img width="1156" alt="image" src="https://user-images.githubusercontent.com/26592486/183145630-af78c8aa-d7f6-43c8-9de9-a57be94f5eec.png">

So, as you can see, my `/direct` function just gets the original `src` URL like I would want/expect. However, my `/default` function always get a URL of `/default`, which is **not** what I want. Why???

### Update

It looks like removing my `default.prerender-config.json` to turn my `default.func` to a regular serverless function instead of a prerender function causes the issue to disappear and everything works as expected. So I guess this isn't actually an issue with routing at all and a problem with how URLs are passed to prerender functions.

**Comment by AaronBeaudoin** [comment, marked as answer]
Turns out that prerender functions are always passed the same URL, but this doesn't seem to be documented anywhere that I can tell. Changing the function to a regular function by removing `default.prerender-config.json` resolved my issue.

Of course, now I have a new issue, because I have to figure out how to forward the full rewrite URL with query string to my prerender function.

**Reply by amyegan** [reply]
[Routes](https://vercel.com/docs/project-configuration#legacy/routes) is a legacy property and might not be able to do exactly what you're trying to accomplish. I would recommend using [redirects](https://vercel.com/docs/project-configuration#project-configuration/redirects) or [rewrites](https://vercel.com/docs/project-configuration#project-configuration/rewrites) if you want to use `vercel.json` to handle it.

You might also want to take a look at [Edge Middleware](https://vercel.com/docs/concepts/functions/edge-middleware) for redirecting or rewriting URLs. That will give you much more control over what happens for each particular route scenario.

**Reply by AaronBeaudoin** [reply]
This is not quite accurate. `routes` is only deprecated if _you're specifying it in `vercel.json`_. I'm not using `vercel.json`, I'm using the [**Build Output API**](https://vercel.com/docs/build-output-api/v3#build-output-configuration/supported-properties/routes), and specifying my routes in `.vercel/output/config.json`.

The docs have this to say:

> The `routes` property describes the routing rules that will be applied to the Deployment. It uses the same syntax as the [`routes` property of the `vercel.json` file](https://vercel.com/docs/project-configuration#legacy/routes).

There is no mention of them be deprecated anywhere for usage with the Build Output API, which itself doesn't even support the higher level config of `vercel.json`. You can, however, transform the config yourself as described [here](https://vercel.com/docs/build-output-api/v3#features/high-level-routing). The linked section clarifies:

> The `vercel.json` file supports an [easier-to-use syntax for routing through properties like `rewrites`, `headers`, etc](https://vercel.com/docs/project-configuration). However, the [`config.json` "routes" property](https://vercel.com/docs/build-output-api/v3#build-output-configuration/routes) supports a lower-level syntax.

So, to be clear, the `routes` property is my **only** option in the Build Output API config file without transforming a config myself, in which case why wouldn't I just manually write the same config my transformation would output? And being a lower-level API, if any config property offered by Vercel should be able to do what I want, it's definitely the `routes` property.

Concerning edge middleware, I know that they offer more control over routing, but as I discovered above (and by @TooTallNate's comment below), the issue here isn't routing at all, it's the fact that a prerender function doesn't get the unmodified original URL during a rewrite like a regular serverless function does.

**Reply by amyegan** [reply]
There are, apparently, still some things I need to learn about the Build Output API. I'm very sorry you ran into this problem, and I really appreciate you coming back to share this information so I (and others finding this later) could learn from it.

**Comment by TooTallNate** [comment]
Hi @AaronBeaudoin. Please take a look at [this Prerender function example](https://github.com/vercel/examples/blob/main/build-output-api/prerender-functions/.vercel/output/functions/blog/post.func/index.js#L13), where the `x-now-route-matches` header is checked rather then `req.url`. I believe this will contain the information that you are looking for.

I do agree though that this is inconsistent behavior compared to how a "regular" Serverless Function behaves, and I would like to have that discrepancy be fixed. We are looking into it.

**Reply by AaronBeaudoin** [reply]
Thank you very much for the response. I did look into that example after writing this question, and discovered exactly what you describe. The discrepancy between how a prerender and serverless function behave is indeed unfortunate, but the really killer thing that made my life difficult is the fact that **it is not documented anywhere** apart from that example. And the example doesn't really make sense fully until you go ahead and try doing it yourself, because the docs say this:

> A Prerender asset is a Serverless Function that will be cached by the Vercel Edge Network in the same way as a static file.

Reading that, I fully expected the prerender function to behave exactly the same way as a serverless function, except that the returned response is cached on the edge until it is revalidated either by a TTL or an on-demand invalidation. I learned that I was wrong only through much experimentation with different paths and route configs.

But there's another glaring limitation here, and that is when using the `x-now-route-matches` header while rewriting to a prerender function you can't pass the original URL's query parameters to the prerender function. The query parameters are inaccessible in the routing config, so they're just completely lost in the rewrite because of how the prerender function doesn't get the original URL.

So in my use case, an ecommerce page like `/product/<name>?color=black&size=large`, I have no way to pass those parameters to the prerender function without instead supplying them in the path itself, in which case now the **order matters**, which is a problem if I have a lot of product attributes and I only want to pass one or two of them. Hopefully you can see the issue here.
